appscale/tools/node_layout.py

```diff
@@ -367,202 +317,259 @@ def is_valid_simple_format(self):
     deployment strategy, and if so, constructs self.nodes from it.
-
+    
     Returns:
-      A dict that indicates if the deployment strategy is valid, and if
-      not, the reason why it is invalid.
+      True if the deployment strategy is valid.
+    Raises:
+      BadConfigurationException with reason if the deployment strategy is not 
+        valid.
     """
-    if self.nodes:
-      return self.valid()
-
+    # Keep track of whether the deployment is valid while going through.
     node_hash = {}
+    role_count = {
+      'appengine': 0,
+      'master': 0,
+      'memcache': 0,
+      'taskqueue': 0,
+      'zookeeper': 0,
+      'login': 0
+    }
+    node_count = 1
+    db_master_created = False
+    tq_master_created = False
+    master_node = None
+    login_found = False
+    # Loop through the list of "node sets", which are grouped by role.
     for role, ips in self.input_yaml.iteritems():
-      
-      if isinstance(ips, str):
-        ips = [ips]
+      if len(ips) == 0:
+        self.invalid("Node amount cannot be zero.")
+      ips_list = ips if isinstance(ips, list) else [ips]
+      test_ips = [self.is_cloud_ip(ip) for ip in ips_list]
+      cloud = all(test_ips) or not any(test_ips)
 
-      for index, ip_addr in enumerate(ips):
-        node = None
-        if ip_addr in node_hash:
-          node = node_hash[ip_addr]
-        else:
-          ip_addr, cloud = self.parse_ip(ip_addr)
-          node = AdvancedNode(ip_addr, cloud)
+      # Immediately fail if we have more than one node for master.
+      if role == 'master' and master_node is not None:
+        self.invalid("Only one master is allowed")
 
-        if role == 'database':
-          # The first database node is the master
-          if index == 0:
-            is_master = True
-          else:
-            is_master = False
-          node.add_db_role(is_master)
-        elif role == 'db_master':
-          node.add_role('zookeeper')
-          node.add_role('role')
-        elif role == 'taskqueue':
-          # Like the database, the first taskqueue node is the master
-          if index == 0:
-            is_master = True
-          else:
-            is_master = False
-          node.add_role('taskqueue')
-          node.add_taskqueue_role(is_master)
-        else:
-          node.add_role(role)
-        
-        node_hash[ip_addr] = node
+      # Create or retrieve the nodes from the node_hash.
+      nodes = [node_hash[ip] if ip in node_hash else AdvancedNode(ip, cloud)
+               for ip in ips_list]
 
-    # Dont need the hash any more, make a nodes list
-    nodes = node_hash.values()
+      # Validate volume usage, there should be an equal number of volumes to
+      # number of nodes.
+      if self.disks:
+        self.validate_disks(nodes)
 
-    for node in nodes:
-      if not node.is_valid():
-        return self.invalid(",".join(node.errors()))
+        for node in nodes:
+          node.disk = self.disks.get(node.public_ip)
```

Comment:
This case is going to be deprecated soon. Additionally this is the current `node-1: disk-1` where the public ip actually is the key.